src/Control/FileRemapper.php

```diff
@@ -0,0 +1,274 @@
+<?php
+
+namespace SilverStripe\Control;
+
+use InvalidArgumentException;
+use RecursiveDirectoryIterator;
+use RecursiveIteratorIterator;
+use SilverStripe\Assets\Filesystem;
+use SilverStripe\Core\Config\Config;
+use SilverStripe\Core\Flushable;
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\Core\Manifest\Module;
+use SilverStripe\Core\Manifest\ModuleLoader;
+use SilverStripe\Core\Manifest\ModuleResource;
+
+/**
+ * Assists with exposing resources from vendor folder to public
+ */
+class FileRemapper implements Flushable
+{
+    /**
+     * Remap with symlink
+     */
+    const SYMLINK = 'symlink';
+
+    /**
+     * Remap with copy
+     */
+    const COPY = 'copy';
+
+    /**
+     * Don't remap (rely on url rewriting only)
+     */
+    const NONE = 'none';
+
+    /**
+     * Remap methods in order of preference
+     *
+     * @var array
+     */
+    private static $methods = [
+        self::SYMLINK,
+        self::COPY
+    ];
+
+    /**
+     * Rewrites applied to relative path before being turned into url
+     *
+     * @config
+     * @var array
+     */
+    private static $path_rewrites = [
+        '#^vendor/#i' => 'resources/',
+    ];
+
+    /**
+     * @param string $relativePath Relative path to rewrite
+     * @param ModuleResource $resource Optional module resource
+     * @param bool $remapPath Enable just-in-time path remapping
+     * @return string
+     */
+    public function rewritePath($relativePath, ModuleResource $resource = null, $remapPath = true)
+    {
+        // Apply path rewrites
+        $rules = Config::inst()->get(static::class, 'path_rewrites') ?: [];
+        $targetPath = $relativePath;
+        foreach ($rules as $from => $to) {
+            $targetPath = preg_replace($from, $to, $targetPath);
+        }
+
+        // JIT Remap if not available
+        if ($remapPath && !file_exists(Director::baseFolder() . '/' . $targetPath)) {
+            $this->remapPath($relativePath, $resource);
+        }
+
+        return $targetPath;
+    }
+
+    /**
+     * Remap all modules
+     */
+    public function remapAllModules()
+    {
+        if (!$this->hasRemapActions()) {
+            return;
+        }
+        foreach (ModuleLoader::inst()->getManifest()->getModules() as $module) {
+            $this->remapModule($module);
+        }
+    }
+
+    /**
+     * Setup paths for the given module
+     *
+     * @param Module $module
+     */
+    protected function remapModule(Module $module)
+    {
+        if (!$this->hasRemapActions()) {
+            return;
+        }
+        // Get list of public folders to expose
+        $exposed = $module->getExposedFolders();
```

Comment:
IMO module should expose `getComposerData()` or `getComposerExtraData($key)`. Let this class take responsibility for deciding where in the composer file the settings are kept.